File: ios/OtaPreferences.swift
import Foundation

/// Local persistence for the OTA system. The safe-mode flags mirror the Remote
/// Config kill switches so the NEXT launch stays on the asset bundle even offline.
enum OtaPreferences {

  private static let safeModeKey = "ota_safe_mode_enabled"
  private static let appVersionKey = "ota_app_version"
  // Prefixed so a module configKey can never collide with the global key.
  private static let moduleSafeModePrefix = "module_safe_mode_"

  static var isSafeModeEnabled: Bool {
    UserDefaults.standard.bool(forKey: safeModeKey)
  }

  static func setSafeModeEnabled(_ enabled: Bool) {
    UserDefaults.standard.set(enabled, forKey: safeModeKey)
  }

  static func isModuleSafeModeEnabled(_ configKey: String) -> Bool {
    UserDefaults.standard.bool(forKey: moduleSafeModePrefix + configKey)
  }

  static func setModuleSafeModeEnabled(_ configKey: String, _ enabled: Bool) {
    UserDefaults.standard.set(enabled, forKey: moduleSafeModePrefix + configKey)
  }

  static var storedAppVersion: String? {
    UserDefaults.standard.string(forKey: appVersionKey)
  }

  static func setStoredAppVersion(_ version: String) {
    UserDefaults.standard.set(version, forKey: appVersionKey)
  }
}

File: android/src/main/java/in/cashify/otaupdate/OtaPreferences.kt
package `in`.cashify.otaupdate

import android.content.Context

object OtaPreferences {

    private const val PREFS_NAME = "cashify_ota_prefs"
    private const val KEY_SAFE_MODE = "safe_mode_enabled"
    private const val KEY_APP_VERSION = "app_version"

    // Prefixed so a module configKey can never collide with the global
    // safe-mode key (e.g. configKey "enabled").
    private const val KEY_MODULE_SAFE_MODE_PREFIX = "module_safe_mode_"

    fun isSafeModeEnabled(context: Context): Boolean {
        return context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .getBoolean(KEY_SAFE_MODE, false)
    }

    fun setSafeModeEnabled(context: Context, enabled: Boolean) {
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .edit().putBoolean(KEY_SAFE_MODE, enabled).apply()
    }

    fun isModuleSafeModeEnabled(context: Context, configKey: String): Boolean {
        return context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .getBoolean(KEY_MODULE_SAFE_MODE_PREFIX + configKey, false)
    }

    fun setModuleSafeModeEnabled(context: Context, configKey: String, enabled: Boolean) {
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .edit().putBoolean(KEY_MODULE_SAFE_MODE_PREFIX + configKey, enabled).apply()
    }

    fun getStoredAppVersion(context: Context): String? {
        return context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .getString(KEY_APP_VERSION, null)
    }

    fun setStoredAppVersion(context: Context, version: String) {
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .edit().putString(KEY_APP_VERSION, version).apply()
    }
}
